File: sdk/swift/Sources/ConduytKit/Modules/OLED.swift
import Foundation

/// Typed wrapper for the CONDUYT SSD1306 OLED module.
@available(iOS 15.0, macOS 12.0, *)
public class ConduytOLED {
    private let device: ConduytDevice
    private let moduleID: UInt8

    public init(device: ConduytDevice, moduleID: UInt8) {
        self.device = device
        self.moduleID = moduleID
    }

    /// Initialize the display. addr=0 falls back to 0x3C on the firmware.
    public func begin(width: UInt8 = 128, height: UInt8 = 64, addr: UInt8 = 0x3C) async throws {
        _ = try await device.modCmd(Data([moduleID, 0x01, width, height, addr]))
    }

    /// Clear the off-screen buffer. Call `show()` to flush.
    public func clear() async throws {
        _ = try await device.modCmd(Data([moduleID, 0x02]))
    }

    /// Render text into the off-screen buffer.
    public func text(x: UInt8, y: UInt8, size: UInt8, _ s: String) async throws {
        var p = Data([moduleID, 0x03, x, y, size])
        p.append(contentsOf: Array(s.utf8))
        _ = try await device.modCmd(p)
    }

    /// Draw a rectangle. fill=true for a filled rect.
    public func drawRect(x: UInt8, y: UInt8, w: UInt8, h: UInt8, fill: Bool = false) async throws {
        _ = try await device.modCmd(Data([moduleID, 0x04, x, y, w, h, fill ? 1 : 0]))
    }

    /// Draw a 1-bit bitmap. data must hold ceil(w*h/8) bytes.
    public func drawBitmap(x: UInt8, y: UInt8, w: UInt8, h: UInt8, data: Data) async throws {
        var p = Data([moduleID, 0x05, x, y, w, h])
        p.append(data)
        _ = try await device.modCmd(p)
    }

    /// Flush the off-screen buffer to the panel.
    public func show() async throws {
        _ = try await device.modCmd(Data([moduleID, 0x06]))
    }
}

File: sdk/rust/src/modules/oled.rs
//! CONDUYT OLED SSD1306 Module — typed wrapper

use crate::device::{Device, DeviceError};
use crate::transports::Transport;
extern crate alloc;
use alloc::vec::Vec;

/// SSD1306 OLED module wrapper.
pub struct OLED<'a, T: Transport> {
    device: &'a mut Device<T>,
    module_id: u8,
}

impl<'a, T: Transport> OLED<'a, T> {
    pub fn new(device: &'a mut Device<T>, module_id: u8) -> Self {
        Self { device, module_id }
    }

    /// Initialize the display. `addr = 0` falls back to 0x3C on the firmware.
    pub fn begin(&mut self, width: u8, height: u8, addr: u8) -> Result<(), DeviceError<T::Error>> {
        self.device.mod_cmd(&[self.module_id, 0x01, width, height, addr])?;
        Ok(())
    }

    /// Clear the off-screen buffer. Call `show` to flush.
    pub fn clear(&mut self) -> Result<(), DeviceError<T::Error>> {
        self.device.mod_cmd(&[self.module_id, 0x02])?;
        Ok(())
    }

    /// Render text into the buffer.
    pub fn text(&mut self, x: u8, y: u8, size: u8, s: &str) -> Result<(), DeviceError<T::Error>> {
        let bytes = s.as_bytes();
        let mut p: Vec<u8> = Vec::with_capacity(5 + bytes.len());
        p.push(self.module_id);
        p.push(0x03);
        p.push(x);
        p.push(y);
        p.push(size);
        p.extend_from_slice(bytes);
        self.device.mod_cmd(&p)?;
        Ok(())
    }

    /// Draw a rectangle. `fill = true` for a filled rect.
    pub fn draw_rect(&mut self, x: u8, y: u8, w: u8, h: u8, fill: bool) -> Result<(), DeviceError<T::Error>> {
        self.device.mod_cmd(&[self.module_id, 0x04, x, y, w, h, fill as u8])?;
        Ok(())
    }

    /// Draw a 1-bit bitmap. `data` must hold ceil(w*h/8) bytes.
    pub fn draw_bitmap(&mut self, x: u8, y: u8, w: u8, h: u8, data: &[u8]) -> Result<(), DeviceError<T::Error>> {
        let mut p: Vec<u8> = Vec::with_capacity(6 + data.len());
        p.push(self.module_id);
        p.push(0x05);
        p.push(x);
        p.push(y);
        p.push(w);
        p.push(h);
        p.extend_from_slice(data);
        self.device.mod_cmd(&p)?;
        Ok(())
    }

    /// Flush the off-screen buffer to the panel.
    pub fn show(&mut self) -> Result<(), DeviceError<T::Error>> {
        self.device.mod_cmd(&[self.module_id, 0x06])?;
        Ok(())
    }
}
